app_dart/lib/src/request_handlers/check_for_waiting_pull_requests.dart

```diff
@@ -0,0 +1,280 @@
+    await _checkPRs('flutter', 'engine', log, client);
+
+    return Body.empty;
+  }
+
+  Future<void> _checkPRs(
+    String owner,
+    String name,
+    Logging log,
+    GraphQLClient client,
+  ) async {
+    bool hasMerged = false;
+    final Map<String, dynamic> data = await _queryGraphQL(
+      owner,
+      name,
+      log,
+      client,
+    );
+    for (_AutoMergeQueryResult queryResult in _parseQueryData(data)) {
+      if (!hasMerged && queryResult.shouldMerge) {
+        hasMerged = await _mergePullRequest(
+          queryResult.graphQLId,
+          queryResult.sha,
+          log,
+          client,
+        );
+      } else if (queryResult.shouldRemoveLabel) {
+        await _removeLabel(
+          queryResult.graphQLId,
+          queryResult.removalMessage,
+          queryResult.labelId,
+          client,
+        );
+      }
+    }
+  }
+
+  Future<Map<String, dynamic>> _queryGraphQL(
+    String owner,
+    String name,
+    Logging log,
+    GraphQLClient client,
+  ) async {
+    final String labelName = await config.waitingForTreeToGoGreenLabelName;
+
+    final QueryResult result = await client.query(
+      QueryOptions(
+        document: labeledPullRequestsWithReviewsQuery,
+        fetchPolicy: FetchPolicy.noCache,
+        variables: <String, dynamic>{
+          'sOwner': owner,
+          'sName': name,
+          'sLabelName': labelName,
+        },
+      ),
+    );
+
+    if (result.hasErrors) {
+      log.error(jsonEncode(result.errors));
+      throw const BadRequestException('GraphQL query failed');
+    }
+
+    return result.data;
+  }
+
+  Future<bool> _removeLabel(
+    String id,
+    String message,
+    String labelId,
+    GraphQLClient client,
+  ) async {
+    final QueryResult result = await client.mutate(MutationOptions(
+      document: removeLabelMutation,
+      variables: <String, dynamic>{
+        'id': id,
+        'sBody': message,
+        'labelId': labelId,
+      },
+    ));
+    if (result.hasErrors) {
+      log.error(jsonEncode(result.errors));
+      return false;
+    }
+    return true;
+  }
+
+  Future<bool> _mergePullRequest(
+    String id,
+    String sha,
+    Logging log,
+    GraphQLClient client,
+  ) async {
+    final QueryResult result = await client.mutate(MutationOptions(
+      document: mergePullRequestMutation,
+      variables: <String, dynamic>{
+        'id': id,
+        'oid': sha,
+      },
+    ));
+
+    if (result.hasErrors) {
+      log.error(jsonEncode(result.errors));
+      return false;
+    }
+    return true;
+  }
+
+  /// Parses a GraphQL query to a list of [_AutoMergeQueryResult]s.
+  ///
+  /// This method will not return null, but may return an empty list.
+  List<_AutoMergeQueryResult> _parseQueryData(Map<String, dynamic> data) {
+    final Map<String, dynamic> repository = data['repository'];
+    if (repository == null || repository.isEmpty) {
+      return <_AutoMergeQueryResult>[];
```

Comment:
This sounds like an exception.

Reply:
Hm.  Probably a good idea rather than returning an empty list - I believe the schema for this allows null, but that would indicate we asked for a the wrong repo (or we don't have access to it I guess?).

Reply:
Yeah, null or empty repository feels like an error in our configuration.